Add the output reference page and align with the other block titles (#36934)
hashicorp/terraform: website/docs/language/block/output.mdx

---
page_title: Output block reference for the Terraform configuration language
description: Use the `output` block to expose information about your module's infrastructure.
---

# `output` block reference

The `output` block lets you expose information about your infrastructure.

> **Hands-on:** Try the [Output Data From Terraform](/terraform/tutorials/configuration-language/outputs) tutorial.

## Background

> The value of an `output` block is similar to a return value in other programming languages. The `output` block exposes information about your infrastructure that you can reference on the command line, in HCP Terraform, and in other Terraform configurations that can access your configuration's state. The `output` block serves three main purposes in Terraform:

Author: Ooooh great point! I added `- Passing information from a Terraform operation to an automation tool.` as a bullet point here to capture this. I think I'll leave off the `-json` portion since that's more related to the CLI, but thank you for calling this out!

```hcl
output "unique_name" {
  value = <expression-to-evaluate>
}
```

Any valid expression can be an output `value.` Refer to [Expressions](/terraform/language/expressions/) to learn more.

#### Summary

- Data type: Expression
- Default: None
- Required: Yes

### `description`

Use the `description` argument in an `output` block to explain the purpose of this output, requirements, and what kind of value this output exports.

```hcl
output "instance_ip_addr" {
  value       = <expression-to-evaluate>
  description = "<concise description from the perspective of a module consumer>."
}
```
We recommend writing the description from the point of view of a module consumer to help consumers understand why the output exists and how to use it.

#### Summary

- Data type: String
- Default: None

### `sensitive`

Add the `sensitive` argument to an `output` block to hide that block's value in the CLI output of a Terraform operation.

```hcl
output "admin_password" {
  value     = <expression-to-evaluate>
  sensitive = <boolean>
}
```

> If you set the `sensitive` argument to `true`, Terraform redacts that output value in plan and apply operations:

Author: Great point! Added a section below to cover this:

```
If you use the [`terraform output` command](/terraform/cli/commands/output) with the `-json` or `-raw` command-line flags, Terraform displays `sensitive` values in plain text.
```

- The value of the `output` block must come from an ephemeral context.
- You can only reference that output in other ephemeral contexts.

The following ephemeral contexts can set values for output blocks with the `ephemeral` argument, and they can also reference output blocks with the `ephemeral` argument:
- Another child module's ephemeral `output` block
- A [write-only argument](/terraform/language/resources/ephemeral/write-only)
- [Variables with the `ephemeral` argument](/terraform/language/values/variables#exclude-values-from-state)
- [An `ephemeral` resource block](/terraform/language/resources/ephemeral)
- Configuring providers in the [`provider` block](/terraform/language/providers)
- In [provisioner](/terraform/language/resources/provisioners/syntax) and [connection](/terraform/language/resources/provisioners/connection) blocks

#### Summary

- Data type: Boolean
- Default: `false`

### `depends_on`

> The `depends_on` argument specifies an explicit dependency on another upstream resource for an `output` block. Terraform must complete all operations on an upstream resource before an `output` block can reference it.

Reviewer: ```suggestion
The `depends_on` argument specifies an explicit dependency on another upstream resource for an `output` block. Terraform will complete all operations on the upstream resource before it computes the value of an `output` block that depends on it.
```

We could also say "all upstream resources ... the values of output blocks that depend on it."

| `error_message` | Message to display if the condition evaluates to `false`. | String | Required |

Terraform evaluates preconditions on outputs when creating or applying a plan, and if a precondition block's `condition` expression evaluates to `false`, then Terraform throws an error with the `error_message` and stops the current operation.

Refer to [Validate your configuration](/terraform/language/test-and-validate/validate#preconditions) for more details about preconditions.

#### Summary

- Data type: Block
- Default: None

## Examples

The following examples demonstrate common use cases for `output` values.

### Basic example

In the following example, an `output` block exposes a server's private IP address from the current module:

```hcl
output "instance_ip_addr" {
  value       = aws_instance.server.private_ip
  description = "The private IP address of the main server instance."
}
```

### Accessing child module outputs

In the following example, the current module uses an output named `instance_ip_addr` from a child module named `web_server`:

```hcl
output "website_url" {
  value = "https://${module.web_server.instance_ip_addr}"
}

> ```

Reviewer: As best practice, I think all of these should have a description field.
description = "The website URL, starting with https://"

### Sensitive output

In the following example, the `db_password` output sets the `sensitive` argument to `true` to prevent Terraform from displaying the database password in CLI output:

```hcl
output "db_password" {
  value     = aws_db_instance.db.password
  sensitive = true

> }

Reviewer: recommend adding description
description = "The database password."

```

### Output with validation

In the following example, the `instance_public_ip` output has a `precondition` to ensure certain conditions are met before exposing the output or saving it to state. Before `instance_public_ip` shares the public IP for server, the output's precondition ensures a server's security group has at least one ingress rule to allow traffic on ports `80` or `443`:

```hcl
output "instance_public_ip" {
  value = aws_instance.web.public_ip

  precondition {
    condition     = length([for rule in aws_security_group.web.ingress : rule if rule.to_port == 80 || rule.to_port == 443]) > 0

>     error_message = "Security group must allow HTTP (port 80) or HTTPS (port 443) traffic."

Reviewer: ```suggestion
    error_message = "Security group must allow HTTP (port 80) or HTTPS (port 443) ingress traffic."
```

  }

> }

Reviewer: description = "Public IP address of the instance."

```
If the server's security group does not allow ingress on ports `80` or `443`, then Terraform throws an error with the `error_message` and stops the current operation.

### Explicit dependencies

In the following example, `instance_ip_addr` adds an explicit dependency on the security group rule to ensure that Terraform creates the security group before exposing the IP address:

```hcl
output "instance_ip_addr" {
  value       = aws_instance.server.private_ip
  description = "The private IP address of the main server instance."

  depends_on = [
    # Services are unreachable unless the security group rule is created
    # before exposing this IP address.
    aws_security_group_rule.local_access,
  ]
}
```

> The `output` block typically does not require explicit dependencies so we recommend adding a comment to explain why it's necessary.

Reviewer: ```suggestion
The `output` block typically does not require explicit dependencies so we recommend that when you add an explicit dependency, include a comment to explain why it's necessary.
```
